### PR TITLE
Implement HP drain

### LINKED ISSUE #5946: Passive HP drain implementation
Health drain is not yet implemented but should be (to match stable).

### PR DESCRIPTION
**Pull request by smoogipoo**
Resolves https://github.com/ppy/osu/issues/5946

Prereqs:
- [x] https://github.com/ppy/osu/pull/7358

Taiko still uses its own "accumulating" health processor, but maybe that'll change in the future too.

This requires the processors to be added to the draw hierarchy. Currently, this is achieved by moving the overlays container of `DrawableRuleset` into the frame stability container and adding the processors to that container. As discussed, a future refactoring will move the processors into `DrawableRuleset` itself (although still be constructed via `Ruleset`).

In terms of rulesets, the breaking change is that their judgement processors don't accept a beatmap via ctor anymore, and is rather given in `ApplyBeatmap()`.

### CODE REVIEW

**Review by bdach — COMMENTED** [review]
I can't find much fault with this after scrolling through, albeit I mostly took the actual drain algorithm as unquestionable gospel since you probably know better what is actually in stable. The only thing I sort of dislike is the `healthIncreases` null/non-null swap in the simulation pass but it's not really anything major.

**Comment by iiSaLMaN** [comment]
I think this closes https://github.com/ppy/osu/issues/5946.

**Comment by smoogipoo** [comment]
@bdach fyi: this doesn't follow stable's algorithm. After discussions we've settled on trying to define HP a bit more formally than stable, as the minimum percentage of HP reached during gameplay for a perfect play.

The concept and execution is similar, however stable only adjusts in one direction - decreasing drain to make it so the player can't ever fail.

Will remove the null/non-null swap, probably best to until we deem it's an issue.

File: osu.Game.Tests/Gameplay/TestSceneDrainingHealthProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+{
+    [HeadlessTest]
+    public class TestSceneDrainingHealthProcessor : OsuTestScene
+    {
+        private Bindable<bool> breakTime;
+        private HealthProcessor processor;
+        private ManualClock clock;
+
+        [Test]
+        public void TestInitialHealthStartsAtOne()
```
Comment (peppy) [inline osu.Game.Tests/Gameplay/TestSceneDrainingHealthProcessor.cs]: Please add tests covering rewind support

**Comment by peppy** [comment]
I changed the drain to start from the first hitobject's starttime rather than `GameplayStartTime` (more closely matches stable). Seems good apart for missing tests.

**Review by peppy — APPROVED** [review]

**Pull request merged**